specreduce/wavelength_calibration.py

```diff
@@ -0,0 +1,248 @@
+                                 f"{self.refinement_method} refinement.")
+        elif self.refinement_method == "gradient" and 'direction' not in self.refinement_kwargs:
+            raise ValueError("You must define 'direction' in refinement_kwargs to use "
+                             "gradient refinement")
+
+    @classmethod
+    def by_name(cls, input_spectrum, line_name, pixel,
+                refinement_method=None, refinement_kwargs={}):
+        # this will do wavelength lookup and passes that wavelength to __init__ allowing the user
+        # to call CalibrationLine.by_name(spectrum, 'H_alpha', pixel=40)
+        # We could also have a type-check for ``wavelength`` above, this just feels more verbose
+        return cls(...)
+
+    def _do_refinement(self, input_spectrum):
+        if self.refinement_method == 'gaussian':
+            window_width = self.refinement_kwargs.get('range')
+            window = SpectralRegion((self.pixel-window_width)*u.pix,
+                                    (self.pixel+window_width)*u.pix)
+
+            input_model = Gaussian1D(mean=self.pixel, stddev=3,
+                                     amplitude=self.input_spectrum.flux[self.pixel])
+
+            fitted_model = fit_lines(self.input_spectrum, input_model, window=window)
+            new_pixel = fitted_model.mean.value
+
+        elif self.refinement_method == 'min':
+            window_width = self.refinement_kwargs.get('range')
+            new_pixel = np.argmin(self.input_spectrum.flux[self.pixel-window_width:
+                                                           self.pixel+window_width+1])
+            new_pixel += self.pixel - window_width
+
+        elif self.refinement_method == 'max':
+            window_width = self.refinement_kwargs.get('range')
+            new_pixel = np.argmax(self.input_spectrum.flux[self.pixel-window_width:
+                                                           self.pixel+window_width+1])
+            new_pixel += self.pixel - window_width
+
+        else:
+            raise ValueError(f"Refinement method {self.refinement_method} is not implemented")
+
+        return new_pixel
+
+    def refine(self, input_spectrum=None, return_object=False):
+        # finds the center of the line according to refinement_method/kwargs and returns
+        # either the pixel value or a CalibrationLine object with pixel changed to the refined value
+        if self.refinement_method is None:
+            return self.pixel
+        input_spectrum = self.input_spectrum if input_spectrum is None else input_spectrum
+        refined_pixel = self._do_refinement(input_spectrum)
+        if return_object:
+            return CalibrationLine(input_spectrum, self.wavelength, refined_pixel,
+                                   refinement_method=self.refinement_method,
+                                   refinement_kwargs=self.refinement_kwargs)
+
+        return refined_pixel
+
+    @cached_property
```

Comment:
should probably be coupled with custom setters for the input spectrum and options to clear the cache